crowbar_framework/app/models/node_object.rb

```diff
@@ -709,21 +709,26 @@ def crowbar_run_list(*args)
   end
 
   def add_to_run_list(rolename, priority, states = nil)
+    Rails.logger.debug("Ensuring #{name} has role #{rolename} with priority #{priority}")
+    save_it = false
     states = ["all"] unless states
-    crowbar["run_list_map"] = {} if crowbar["run_list_map"].nil?
+    crowbar["run_list_map"] ||= {}
     val = { "states" => states, "priority" => priority }
-    crowbar["run_list_map"][rolename] = val
-    Rails.logger.debug("crowbar[run_list_map][#{rolename}] = #{val.inspect}")
-    Rails.logger.debug("current state is #{self.crowbar['state']}")
-
-    # only rebuild the run_list if it effects the current state.
-    self.rebuild_run_list if states.include?("all") or states.include?(self.crowbar["state"])
+    if crowbar["run_list_map"][rolename] != val
+      crowbar["run_list_map"][rolename] = val
+      save_it = true
+    end
+    rebuild_run_list || save_it
   end
 
   def delete_from_run_list(rolename)
-    crowbar["run_list_map"] = {} if crowbar["run_list_map"].nil?
-    crowbar["run_list_map"][rolename] = { "states" => ["all"], "priority" => -1001 } unless crowbar["run_list_map"].nil?
-    crowbar_run_list.run_list_items.delete "role[#{rolename}]"
+    Rails.logger.debug("Ensuring #{name} doesn't have role #{rolename}")
+    crowbar["run_list_map"] ||= {}
+    if crowbar["run_list_map"].key?(rolename)
+      crowbar["run_list_map"].delete(rolename)
```

Comment:
I can't be sure but as I understand this should only add the same roles to run_list and run_list_map so both are consistent with each other. @vuntz what do you think?